### PR TITLE
Add ensure disconnect to ftp login module, ftp timeout configuration

### LINKED ISSUE #17791: auxiliary/scanner/ftp/ftp_login: Sockets are not closed - potentially leading to denial of service
The `auxiliary/scanner/ftp/ftp_login` module uses the `Metasploit::Framework::LoginScanner::FTP` FTP login scanner library, which uses `connect_login` to connect and attempt login for each `username:password` combination.

A new TCP connection is created for each login attempt. The socket is not reused. The socket is never closed. Instead, Metasploit relies on the remote server eventually terminating the connection, which varies depending on the server's timeout configuration.

This causes sockets to hang open, potentially exhausting available sockets in the FTP server client connection pool (if such a a limit is enforced), resulting in denial of service. If no such limit is enforced, a brute-force attack of sufficient size combined with a sufficiently lengthy server timeout would eventually exhaust all available sockets on both the server and client.

```
msf6 auxiliary(scanner/ftp/ftp_login) > run

[*] 192.168.200.158:21    - 192.168.200.158:21 - Starting FTP login sweep
[!] 192.168.200.158:21    - No active DB -- Credential data will not be saved!
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!@#$% (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!@#$%^ (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!@#$%^& (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!@#$%^&* (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!boerbul (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!boerseun (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!gatvol (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!hotnot (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!kak (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!koedoe (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!likable (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!poes (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!pomp (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!soutpiel (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:.net (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:0 (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:000000 (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:00000000 (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:0007 (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:007 (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:007007 (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:0s (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:0th (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:1 (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:10 (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:100 (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:1000 (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:1000s (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:100s (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:1022 (Incorrect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:10s (Unable to Connect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:10sne1 (Unable to Connect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:1111 (Unable to Connect: )
[*] 192.168.200.158:21    - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
rmsf6 auxiliary(scanner/ftp/ftp_login) > run

[*] 192.168.200.158:21    - 192.168.200.158:21 - Starting FTP login sweep
[!] 192.168.200.158:21    - No active DB -- Credential data will not be saved!
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!@#$% (Unable to Connect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!@#$%^ (Unable to Connect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!@#$%^& (Unable to Connect: )
[*] 192.168.200.158:21    - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/ftp/ftp_login) > run

[*] 192.168.200.158:21    - 192.168.200.158:21 - Starting FTP login sweep
[!] 192.168.200.158:21    - No active DB -- Credential data will not be saved!
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!@#$% (Unable to Connect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!@#$%^ (Unable to Connect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!@#$%^& (Unable to Connect: )
[*] 192.168.200.158:21    - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/ftp/ftp_login) > run

[*] 192.168.200.158:21    - 192.168.200.158:21 - Starting FTP login sweep
[!] 192.168.200.158:21    - No active DB -- Credential data will not be saved!
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!@#$% (Unable to Connect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!@#$%^ (Unable to Connect: )
[-] 192.168.200.158:21    - 192.168.200.158:21 - LOGIN FAILED: test:!@#$%^& (Unable to Connect: )
[*] 192.168.200.158:21    - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/ftp/ftp_login) > 
```

Server shows maximum connection limit reached:

```
...
2023-03-19 23:38:50,462 ubuntu-22-04-amd64 proftpd[122519] ubuntu-22-04-amd64 (192.168.200.130[192.168.200.130]): FTP session opened.
2023-03-19 23:38:50,464 ubuntu-22-04-amd64 proftpd[122519] ubuntu-22-04-amd64 (192.168.200.130[192.168.200.130]): USER test: no such user found from 192.168.200.130 [192.168.200.130] to ::ffff:192.168.200.158:21
2023-03-19 23:38:50,503 ubuntu-22-04-amd64 proftpd[122521] ubuntu-22-04-amd64 (192.168.200.130[192.168.200.130]): FTP session opened.
2023-03-19 23:38:50,505 ubuntu-22-04-amd64 proftpd[122521] ubuntu-22-04-amd64 (192.168.200.130[192.168.200.130]): USER test: no such user found from 192.168.200.130 [192.168.200.130] to ::ffff:192.168.200.158:21
2023-03-19 23:38:50,547 ubuntu-22-04-amd64 proftpd[122523] ubuntu-22-04-amd64 (192.168.200.130[192.168.200.130]): FTP session opened.
2023-03-19 23:38:50,551 ubuntu-22-04-amd64 proftpd[122523] ubuntu-22-04-amd64 (192.168.200.130[192.168.200.130]): USER test: no such user found from 192.168.200.130 [192.168.200.130] to ::ffff:192.168.200.158:21
2023-03-19 23:38:50,563 ubuntu-22-04-amd64 proftpd[122072] ubuntu-22-04-amd64: MaxInstances (30) reached, new connection denied
2023-03-19 23:38:50,567 ubuntu-22-04-amd64 proftpd[122072] ubuntu-22-04-amd64: MaxInstances (30) reached, new connection denied
2023-03-19 23:38:50,572 ubuntu-22-04-amd64 proftpd[122072] ubuntu-22-04-amd64: MaxInstances (30) reached, new connection denied
2023-03-19 23:39:03,920 ubuntu-22-04-amd64 proftpd[122072] ubuntu-22-04-amd64: MaxInstances (30) reached, new connection denied
2023-03-19 23:39:03,924 ubuntu-22-04-amd64 proftpd[122072] ubuntu-22-04-amd64: MaxInstances (30) reached, new connection denied
2023-03-19 23:39:03,928 ubuntu-22-04-amd64 proftpd[122072] ubuntu-22-04-amd64: MaxInstances (30) reached, new connection denied
2023-03-19 23:39:12,065 ubuntu-22-04-amd64 proftpd[122072] ubuntu-22-04-amd64: MaxInstances (30) reached, new connection denied
2023-03-19 23:39:12,068 ubuntu-22-04-amd64 proftpd[122072] ubuntu-22-04-amd64: MaxInstances (30) reached, new connection denied
2023-03-19 23:39:12,071 ubuntu-22-04-amd64 proftpd[122072] ubuntu-22-04-amd64: MaxInstances (30) reached, new connection denied
2023-03-19 23:39:36,839 ubuntu-22-04-amd64 proftpd[122072] ubuntu-22-04-amd64: MaxInstances (30) reached, new connection denied
2023-03-19 23:39:36,842 ubuntu-22-04-amd64 proftpd[122072] ubuntu-22-04-amd64: MaxInstances (30) reached, new connection denied
2023-03-19 23:39:36,844 ubuntu-22-04-amd64 proftpd[122072] ubuntu-22-04-amd64: MaxInstances (30) reached, new connection denied
```

Sockets remain open in `ESTABLISHED` state:

```
# netstat -an | grep 192.168.200.158:21
tcp        0      0 192.168.200.130:32813   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:32975   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:41899   192.168.200.158:21      CLOSE_WAIT 
tcp        0      0 192.168.200.130:40451   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:45147   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:45685   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:37755   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:34247   192.168.200.158:21      CLOSE_WAIT 
tcp        0      0 192.168.200.130:35535   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:40169   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:34203   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:36795   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:33271   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:45225   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:43331   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:43609   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:38711   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:44791   192.168.200.158:21      CLOSE_WAIT 
tcp        0      0 192.168.200.130:46611   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:46493   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:34191   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:42375   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:46605   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:39407   192.168.200.158:21      CLOSE_WAIT 
tcp        0      0 192.168.200.130:34865   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:33061   192.168.200.158:21      CLOSE_WAIT 
tcp        0      0 192.168.200.130:35663   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:45427   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:35533   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:42337   192.168.200.158:21      CLOSE_WAIT 
tcp        0      0 192.168.200.130:43947   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:43655   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:37457   192.168.200.158:21      CLOSE_WAIT 
tcp        0      0 192.168.200.130:38283   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:43029   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:45629   192.168.200.158:21      ESTABLISHED
tcp        0      0 192.168.200.130:37735   192.168.200.158:21      ESTABLISHED
```

### PR DESCRIPTION
**Pull request by zgoldman-r7**
Closes https://github.com/rapid7/metasploit-framework/issues/17791

Paired with @adfoster-r7 

Ensures that sockets are being closed when using `ftp_login` module.
Additionally, adds ability to set `ftp_timeout` for this module in options prior to run.

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] run against either netcat or a real server
- [ ] ensure that the server's connection closes after run (via `lsof`, etc)

For debugging we ran against a ncat server:

```
metasploit-framework % ncat -lnvp 2000
\Ncat: Version 7.93 ( https://nmap.org/ncat )
Ncat: Listening on :::2000
Ncat: Listening on 0.0.0.0:2000
```

Run the metasploit module:

```
msf6 auxiliary(scanner/ftp/ftp_login) > time run rhosts=127.0.0.1 rport=2000 username=zgoldman password=pass ftptimeout=3

[*] 127.0.0.1:2000        - 127.0.0.1:2000 - Starting FTP login sweep
[!] 127.0.0.1:2000        - No active DB -- Credential data will not be saved!
[-] 127.0.0.1:2000        - 127.0.0.1:2000 - LOGIN FAILED: zgoldman:pass (Incorrect: )
[*] 127.0.0.1:2000        - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
[+] Command "run rhosts=127.0.0.1 rport=2000 username=zgoldman password=pass ftptimeout=3" completed in 6.047376000002259 seconds
```

You can see Metasploit connected to the ncat bind, and attempted to authenticate:

```
metasploit-framework % ncat -lnvp 2000
\Ncat: Version 7.93 ( https://nmap.org/ncat )
Ncat: Listening on :::2000
Ncat: Listening on 0.0.0.0:2000
Ncat: Connection from 127.0.0.1.
Ncat: Connection from 127.0.0.1:49475.
USER zgoldman
```

### Before

The file descriptors are not closed after the module completes:

```
zgoldman@AUS-MBP-10311 metasploit-framework % lsof -i :2000
COMMAND   PID     USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
ruby    32353 zgoldman   15u  IPv4 0xd34de6bc98106953      0t0  TCP localhost:65320->localhost:callbook (ESTABLISHED)
ncat    41191 zgoldman    5u  IPv4 0xd34de6bc973cc643      0t0  TCP localhost:callbook->localhost:65320 (ESTABLISHED)
```

###  After

The file descriptors are now closed correctly:

```
zgoldman@AUS-MBP-10311 metasploit-framework % lsof -i :2000
zgoldman@AUS-MBP-10311 metasploit-framework % 
```

### CODE REVIEW

File: modules/auxiliary/scanner/ftp/ftp_login.rb [hunk 1]
```diff
@@ -70,6 +70,7 @@ def run_host(ip)
         max_send_size: datastore['TCP::max_send_size'],
         send_delay: datastore['TCP::send_delay'],
         connection_timeout: 30,
```
Comment (bwatters-r7) [inline modules/auxiliary/scanner/ftp/ftp_login.rb:72]: Maybe make this user-mutable

Comment (adfoster-r7) [reply modules/auxiliary/scanner/ftp/ftp_login.rb:72]: `ConnectTimeout` might work here as an existing datastore option to use

Comment (jheysel-r7) [reply modules/auxiliary/scanner/ftp/ftp_login.rb:72]: The default value for `ConnectTimeout` is `10`. Do we want to make a separate configurable timeout datastore value for this module? If so, I can add that in before landing.

**Comment by jheysel-r7** [comment]
Thanks for the fix @zgoldman-r7! Changes look good and testing checks out.

Before: 

```
msf6 auxiliary(scanner/ftp/ftp_login) >  time run rhosts=127.0.0.1 rport=2000 username=zgoldman password=pass ftptimeout=3

[*] 127.0.0.1:2000        - 127.0.0.1:2000 - Starting FTP login sweep
[*] 127.0.0.1:2000        - Connect Timeout is: 30
[!] 127.0.0.1:2000        - No active DB -- Credential data will not be saved!
[-] 127.0.0.1:2000        - 127.0.0.1:2000 - LOGIN FAILED: zgoldman:pass (Incorrect: )
[*] 127.0.0.1:2000        - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
[+] Command "run rhosts=127.0.0.1 rport=2000 username=msfuser password=pass ftptimeout=3" completed in 32.04114499996649 seconds
```

```
➜  ~ ncat -lnvp 2000
Ncat: Version 7.92 ( https://nmap.org/ncat )
Ncat: Listening on :::2000
Ncat: Listening on 0.0.0.0:2000
Ncat: Connection from 127.0.0.1.
Ncat: Connection from 127.0.0.1:61183.
USER msfuser
```

```
➜  ~ lsof -i :2000
COMMAND   PID    USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
ruby    77299 jheysel   14u  IPv4 0x7d755b2e6f8eeb4b      0t0  TCP localhost:61183->localhost:callbook (CLOSED)
```

After: 

```
➜  ~ lsof -i :2000
➜  ~
```

File: modules/auxiliary/scanner/ftp/ftp_login.rb (hunk 1 above)
Comment (bwatters-r7) [reply modules/auxiliary/scanner/ftp/ftp_login.rb:72]: I think that makes sense.

**Comment by adfoster-r7** [comment]
@jheysel-r7 / @zgoldman-r7 Just a heads-up that I've created the PR https://github.com/rapid7/metasploit-framework/pull/17908 to remove the custom `CONNECTION_TIMEOUT` option that was added in this PR, and to instead honor the existing `ConnectTimeout` option from the suggestion thread https://github.com/rapid7/metasploit-framework/pull/17873#discussion_r1165639070

Context: The `Msf::Exploit::Remote::Ftp` mixin already defined `ConnectTimeout` via `Exploit::Remote::Tcp` ([source](https://github.com/rapid7/metasploit-framework/blob/4cb9a4c1421ce08e36cb4da30ae65d7be886c9cb/lib/msf/core/exploit/remote/tcp.rb#L73)), it was just never wired up correctly in this module. Instead of creating a new option, the PR just makes use of the existing option that users expect to be able to configure

**Pull request merged**